File: src/test/java/sirius/kernel/xml/XMLReaderTest.java
/*
 * Made with all the love in the world
 * by scireum in Remshalden, Germany
 *
 * Copyright by scireum GmbH
 * http://www.scireum.de - info@scireum.de
 */

package sirius.kernel.xml;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import sirius.kernel.commons.ValueHolder;
import sirius.kernel.health.Counter;
import sirius.kernel.SiriusExtension;

import java.io.ByteArrayInputStream;
import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

@ExtendWith(SiriusExtension.class)
class XMLReaderTest {

    @Test
    @DisplayName("XMLReader extracts XPATH expression")
    void readXpath() throws Exception {
        ValueHolder<String> readString = ValueHolder.of(null);
        Counter nodeCount = new Counter();
        XMLReader reader = new XMLReader();
        reader.addHandler("test", node -> {
            nodeCount.inc();
            readString.set(node.queryString("value"));
        });

        reader.parse(new ByteArrayInputStream(
                "<doc><test><value>1</value></test><test><value>2</value></test><test><value>5</value></test></doc>".getBytes()));

        assertEquals("5", readString.get());
        assertEquals(3, nodeCount.getCount(), "parsed invalid count of nodes");
    }

    @Test
    @DisplayName("XMLReader supports compound XPATH paths")
    void readXpathCompound() throws Exception {
        ValueHolder<Boolean> shouldToggle = ValueHolder.of(false);
        ValueHolder<Boolean> shouldNotToggle = ValueHolder.of(false);
        XMLReader reader = new XMLReader();
        reader.addHandler("doc/test/value", node -> shouldToggle.set(true));
        reader.addHandler("value", node -> shouldNotToggle.set(true));

        reader.parse(new ByteArrayInputStream(
                "<doc><test><value>1</value></test><test><value>2</value></test><test><value>5</value></test></doc>".getBytes()));

        assertTrue(shouldToggle.get());
        assertFalse(shouldNotToggle.get());
    }

    @Test
    @DisplayName("XMLReader reads attributes")
    void readXpathAttributes() throws Exception {
        Map<String, String> attributes = new HashMap<>();
        ValueHolder<String> attribute = ValueHolder.of("");
        XMLReader reader = new XMLReader();
        reader.addHandler("test", node -> {
            attributes.putAll(node.getAttributes());
            attribute.set(node.getAttribute("namedAttribute").asString());
        });

        reader.parse(new ByteArrayInputStream("<doc><test namedAttribute=\"abc\" namedAttribute2=\"xyz\">1</test></doc>".getBytes()));

        assertEquals(2, attributes.size());
        assertEquals("abc", attribute.get());
    }

    @Test
    @DisplayName("reading non existing attributes does not throw errors")
    void readXpathMissingAttributes() throws Exception {
        Map<String, String> attributes = new HashMap<>();
        ValueHolder<String> attribute = ValueHolder.of("wrongValue");
        XMLReader reader = new XMLReader();
        reader.addHandler("test", node -> {
            attributes.putAll(node.getAttributes());
            attribute.set(node.getAttribute("namedAttribute").asString());
        });

        reader.parse(new ByteArrayInputStream("<doc><test>1</test></doc>".getBytes()));

        assertEquals(0, attributes.size());
        assertEquals("", attribute.get());
    }
}

File: src/test/kotlin/sirius/kernel/xml/XmlReaderTest.kt
/*
 * Made with all the love in the world
 * by scireum in Remshalden, Germany
 *
 * Copyright by scireum GmbH
 * http://www.scireum.de - info@scireum.de
 */

package sirius.kernel.xml

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import sirius.kernel.SiriusExtension
import sirius.kernel.commons.ValueHolder
import sirius.kernel.health.Counter
import java.io.ByteArrayInputStream
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * Tests the [XMLReader] class.
 */
@ExtendWith(SiriusExtension::class)
internal class XmlReaderTest {
    @Test
    fun `XMLReader extracts XPATH expression`() {
        val readString = ValueHolder.of<String?>(null)
        val nodeCount = Counter()
        val reader = XMLReader()
        reader.addHandler("test") { node: StructuredNode ->
            nodeCount.inc()
            readString.set(node.queryString("value"))
        }

        reader.parse(
                ByteArrayInputStream(//language=xml
                        """
                            <doc>
                                <test>
                                    <value>1</value>
                                </test>
                                <test>
                                    <value>2</value>
                                </test>
                                <test>
                                    <value>5</value>
                                </test>
                            </doc>
                        """.trimIndent().toByteArray()
                )
        )

        assertEquals("5", readString.get())
        assertEquals(3, nodeCount.count, "parsed invalid count of nodes")
    }

    @Test
    fun `XMLReader supports compound XPATH paths`() {
        val shouldToggle = ValueHolder.of(false)
        val shouldNotToggle = ValueHolder.of(false)
        val reader = XMLReader()
        reader.addHandler("doc/test/value") { _: StructuredNode? ->
            shouldToggle.set(
                    true
            )
        }
        reader.addHandler("value") { _: StructuredNode? -> shouldNotToggle.set(true) }

        reader.parse(
                ByteArrayInputStream(//language=xml
                        """
                            <doc>
                                <test>
                                    <value>1</value>
                                </test>
                                <test>
                                    <value>2</value>
                                </test>
                                <test>
                                    <value>5</value>
                                </test>
                            </doc>
                        """.trimIndent().toByteArray()
                )
        )

        assertTrue { shouldToggle.get() }
        assertFalse { shouldNotToggle.get() }
    }

    @Test
    fun `XMLReader reads attributes`() {
        val attributes: MutableMap<String, String> = HashMap()
        val attribute = ValueHolder.of("")
        val reader = XMLReader()
        reader.addHandler("test") { node: StructuredNode ->
            attributes.putAll(node.getAttributes())
            attribute.set(node.getAttribute("namedAttribute").asString())
        }

        reader.parse(
                ByteArrayInputStream(//language=xml
                        """
                            <doc>
                                <test namedAttribute="abc" namedAttribute2="xyz">1</test>
                            </doc>
                        """.trimIndent().toByteArray()
                )
        )

        reader.parse(ByteArrayInputStream("<doc><test namedAttribute=\"abc\" namedAttribute2=\"xyz\">1</test></doc>".toByteArray()))

        assertEquals(2, attributes.size)
        assertEquals("abc", attribute.get())
    }

    @Test
    fun `Reading non existing attributes does not throw errors`() {
        val attributes: MutableMap<String, String> = HashMap()
        val attribute = ValueHolder.of("wrongValue")
        val reader = XMLReader()
        reader.addHandler("test") { node: StructuredNode ->
            attributes.putAll(node.getAttributes())
            attribute.set(node.getAttribute("namedAttribute").asString())
        }

        reader.parse(
                ByteArrayInputStream(//language=xml
                        """
                            <doc>
                                <test>1</test>
                            </doc>
                        """.trimIndent().toByteArray()
                )
        )

        assertEquals(0, attributes.size)
        assertEquals("", attribute.get())
    }
}
